auth/env_keychain.go

```diff
@@ -12,15 +12,30 @@ import (
 	"github.com/pkg/errors"
 )
 
-// NewKeychain returns either a EnvKeychain or a authn.DefaultKeychain depending on whether the provided environment variable is set
-func NewKeychain(envVar string) authn.Keychain {
+// NewKeychain returns an EnvKeychain based on the provided environment variable.
+// If the provided environment variable is not set, NewKeychain will build an auth config
+// from authn.DefaultKeychain and set the variable in the environment.
+// If any errors are encountered, NewKeychain will return an anonymous keychain.
+func NewKeychain(envVar string, images ...string) authn.Keychain {
 	_, ok := os.LookupEnv(envVar)
 	if !ok {
-		return authn.DefaultKeychain
+		authConfig, err := BuildEnvVar(authn.DefaultKeychain, images...)
+		if err != nil {
+			return &AnonymousKeychain{}
+		}
+		if err = os.Setenv(envVar, authConfig); err != nil {
+			return &AnonymousKeychain{}
+		}
```

Comment:
> putting credentials in the environment when the user doesn't expect it could create security risks
Agree on this point. How bad is it really (just wondering...)?

We initially thought about doing it this way but it requires a lot more changes to the existing code as we'll now need `analyzeCmd`, etc. to hold on to a `keychain` if I'm not mistaken.

Reply:
> How bad is it really (just wondering...)?

As part of lifecycle binaries like `analyzer`, `exporter`, and `creator`? Not that bad b/c we are in control over how the environment is used by the lifecycle and we explicitly strip `CNB_REGISTRY_AUTH` from the buildpack environment when `creator` is used.

However, `NewKeychain` is part of the library interface meaning that folks could be pulling in this code in unpredictable situations. This behavior would likely be surprising to any potential Consumers of `NewKeychain` (I know I would be surprised) and may open folks up to vulnerabilities.

Reply:
Ah, then our latest commit (fff589e) may be helpful (in response to this [comment](https://github.com/buildpacks/lifecycle/pull/409#discussion_r481211308)). We ultimately realized that changing the library method's behavior was not a good idea. Do you think the latest incarnation is okay? Or should we proceed with re-working this to store the auth in memory?

Reply:
TBH setting the env var does bother me a little. However this change goes a long way towards addressing any concrete problems I can foresee arising. Because I agree the changing the library method behavior in a patch is probably a bad idea I think this solution makes sense for `v0.9.2` as is.

However, before we merge this fix into v0.10.0 I think we should:

1. Make `NewKeychain` return a `MultiKeychain` where
   * The `EnvKeychain` or a refactored equivalent is the first keychain
  * A keychain that stores pre-resolved auth-headers from the docker config.json in memory is the second keychain

Reply:
@ekcasey @natalieparellano, I think we addressed all the comments here so I mark this conversation as resolved.
Please reopen if I missed something.